proposals/0540-jsem.rst

```diff
@@ -130,59 +130,86 @@ GHC Jobserver Protocol
-This is achieved by coordination over a system semaphore (either a POSIX
-semaphore [6]_  in the case of Linux and Darwin, or a Win32 semaphore [7]_
-in the case of Windows platforms).
+This is achieved by coordination over a semaphore implementation provided
+by the ``semaphore-compat`` Haskell library.
 
 There are two kinds of participants in the GHC Jobserver protocol:
 
-- The *jobserver* creates a new system semaphore with a certain number of
-  available tokens.
+- The *jobserver* instantiates a new semaphore with a certain number of
+  available tokens. This semaphore is identified by a string, called its
+  **semaphore identifier**
 
   Each time the jobserver wants to spawn a new jobclient subprocess, it **must**
   first acquire a single token from the semaphore, before spawning the subprocess.
   This token **must** be released once the subprocess terminates.
 
-  When spawning the subprocess, the jobserver **should** pass on the name of the
+  When spawning the subprocess, the jobserver **should** pass on the identifier of the
   semaphore tha the jobserver created to the spawned subprocess, in order to
   allow the subprocess to request tokens from the semaphore.
 
   Once work is finished, the jobserver **must** destroy the semaphore it created.
+  At this point, the semaphore identifier is not valid for any operations and
+  attempting to use it will produce an exception.
 
 - A *jobclient* is a subprocess spawned by the jobserver or another jobclient.
 
+  Each jobclient can interact with the semaphore by using a valid semaphore
+  identifier provided to it by a jobserver.
+
   Each jobclient starts with one available token (its *implicit token*,
   which was acquired by the parent which spawned it), and can request more
   tokens through the Jobserver Protocol by waiting on the semaphore.
 
   Each time a jobclient wants to spawn a new jobclient subprocess, it **must**
   pass on a single token to the child jobclient. This token can either be the
   jobclient's implicit token, or another token which the jobclient acquired
-  from the semaphore. It **should** also pass on the name of the semaphore, so
+  from the semaphore. It **should** also pass on the identifier of the semaphore, so
   that the child jobclient may acquire additional tokens from the semaphore.
 
   Each jobclient **must** release exactly as many tokens as it has acquired from
   the semaphore (this does not include the implicit token).
 
-All communication between the jobserver and the jobclient happens through the
-semaphore. This ensures modularity of the protocol: jobclients don't need to
-know anything about the jobserver that has spawned them.
+The **semaphore identifier** must begin a a prefix ``v<N>-`` where the ``<N>``
+specifies the ``semaphore-compat`` major version that this semaphore was created
+using. For instance, a semaphore created with ``semaphore-compat-2.0.0`` could
+be called ``v2-some-semaphore12345``.
```

Comment:
```suggestion
be called ``v2.0-some-semaphore12345``.
```

`2.0` is the major version according to the [PVP](https://pvp.haskell.org/). Of course you could choose to always bump the first component here, but I think it would not hurt to clarify.

Reply:
I think it's fine, maybe even good, to distinguish between the *binary* compatibility of the semaphore implementation and the *source* compatibility of `semaphore-compat`. PVP is about source compatibility, but the authors of `semaphore-compat` might need to make a breaking API change while still preserving binary compatibility with previous versions of the library.

Using the first component of the package version to indicate binary compatibility level seems reasonable to me.